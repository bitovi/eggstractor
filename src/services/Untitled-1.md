source: https://github.com/bitovi/ai-enablement-prompts/tree/main/writing-code/generate-feature

Entered prompt as listed without any changes

```
1. Open the repository on GitHub: https://github.com/bitovi/ai-enablement-prompts.
2. Execute the prompt `writing-code/generate-feature/generate-feature.md`
```

## EGG-46 Implementation Complete ✅

Successfully implemented the "Nested components can be exported as component references" feature following the AI enablement prompt workflow.

### Summary of Changes:

#### 1. **Core Functionality** (`src/services/collection.service.ts`)
- Added `shouldExportAsComponentReference()` - determines when instances should be referenced vs expanded
- Added `createComponentReferenceToken()` - generates component reference tokens
- Added `processComponentReferences()` - second-pass processing for component references
- Modified collection logic to support two-phase processing

#### 2. **SCSS Transformer Updates** (`src/transformers/scss.transformer.ts`)
- Updated `getMixinPropertyAndValue()` to handle component-reference tokens
- Added component reference section in SCSS output
- Generates `@include ComponentName--variant()` statements

#### 3. **Test Coverage** (`src/tests/services/collection.service.test.ts`)
- Added comprehensive tests for component reference functionality
- Tests cover component set instances, reused standalone components, and remote component exclusion
- Validates proper token generation with variant information

#### 4. **Documentation Updates**
- Updated `docs/DOCS.md` with detailed feature explanation
- Updated `README.md` with feature highlights and examples
- Added technical implementation details

### Key Features Implemented:

✅ **Identifies reused components** - Button components reused in modals, etc.
✅ **Handles component variants** - Large outlined button → `@include Button--size-large--theme-primary()`
✅ **Generates SCSS mixins** - Component references as `@include` statements
✅ **Reduces output duplication** - References existing mixins instead of expanding styles
✅ **Remote component exclusion** - Skips components that can't be reliably referenced
✅ **Two-phase processing** - Collects all instances first, then determines references

### Example Output:

```scss
// Component References
// Component reference: Button--size-large--theme-primary
@include Button--size-large--theme-primary()

// Component reference: Modal--variant-compact  
@include Modal--variant-compact()
```

### Next Steps (Out of Scope):
- Handling styling overrides on component instances
- Advanced component nesting detection
- UI controls for enabling/disabling component references

The implementation successfully addresses the core requirements from the Jira ticket and provides a solid foundation for component-based SCSS generation in Eggstractor.

Provided `EGG-46` when prompted for TICKET_NUMBER