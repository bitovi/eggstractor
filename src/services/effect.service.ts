import { VariableToken, TokenCollection } from '../types';
import { sanitizeName } from '../utils';

/**
 * Convert a Figma EffectStyle to a VariableToken
 */
function createEffectStyleToken(effectStyle: EffectStyle): VariableToken | null {
  try {
    // Convert effects array to CSS box-shadow value
    const boxShadowValues = effectStyle.effects
      .filter((effect) => effect.type === 'DROP_SHADOW' || effect.type === 'INNER_SHADOW')
      .map((effect) => {
        const { offset, radius, color, spread } = effect;
        const isInset = effect.type === 'INNER_SHADOW';

        // Convert RGBA to CSS rgba()
        const colorValue = `rgba(${Math.round(color.r * 255)}, ${Math.round(color.g * 255)}, ${Math.round(color.b * 255)}, ${color.a ?? 1})`;

        // Build box-shadow value: [inset] offset-x offset-y blur-radius [spread-radius] color
        const parts = [];
        if (isInset) parts.push('inset');
        parts.push(`${offset?.x || 0}px`);
        parts.push(`${offset?.y || 0}px`);
        parts.push(`${radius || 0}px`);
        if (spread) parts.push(`${spread}px`);
        parts.push(colorValue);

        return parts.join(' ');
      });

    if (boxShadowValues.length === 0) {
      return null;
    }

    const rawValue = boxShadowValues.join(', ');

    return {
      type: 'variable',
      path: [{ name: effectStyle.name, type: 'FRAME' }],
      property: 'box-shadow',
      name: sanitizeName(effectStyle.name),
      value: `$${sanitizeName(effectStyle.name)}`,
      rawValue,
      valueType: null,
      metadata: {
        figmaId: effectStyle.id,
        variableName: effectStyle.name,
        variableTokenType: 'primitive',
      },
    };
  } catch (error) {
    console.warn(`Failed to create token for effect style ${effectStyle.name}:`, error);
    return null;
  }
}

/**
 * Collect all Figma Effect Styles (box-shadow, etc.)
 */
export async function collectAllFigmaEffectStyles(
  collection: TokenCollection,
  onProgress: (progress: number, message: string) => void,
) {
  try {
    onProgress(7, 'Collecting Figma Effect Styles...');
    const effectTokens: VariableToken[] = [];

    // Get all local effect styles
    const effectStyles = await figma.getLocalEffectStylesAsync();

    for (const effectStyle of effectStyles) {
      onProgress(8, `Processing effect style: ${effectStyle.name}`);

      const token = createEffectStyleToken(effectStyle);
      if (token) {
        effectTokens.push(token);
      }
    }

    console.log(`âœ¨ Collected ${effectTokens.length} effect style tokens`);
    collection.tokens.push(...effectTokens);
  } catch (error) {
    console.warn('Failed to collect Figma Effect Styles:', error);
  }
}
