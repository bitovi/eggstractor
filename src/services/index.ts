export * from './collection.service';
export * from './token.service';
export * from './variable.service'; 