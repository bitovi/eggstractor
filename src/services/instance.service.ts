import { StyleToken, InstanceToken, ComponentToken, TokenCollection } from '../types';
import { getProcessorsForNode } from '../processors';
import { extractNodeToken } from './token.service';
import { getNodePathNames } from '../utils/node.utils';

export interface InstanceOverride {
  instanceId: string;
  instanceName: string;
  componentId?: string;
  overrideTokens: StyleToken[];
  allTokens: StyleToken[];
}

/**
 * Detects and extracts override styles for component instances by comparing
 * fresh style extractions from both instance and component nodes.
 */
export async function detectInstanceOverrides(
  collection: TokenCollection
): Promise<InstanceOverride[]> {
  const overrides: InstanceOverride[] = [];

  for (const [instanceId, instanceToken] of Object.entries(collection.instances)) {
    if (instanceToken.remote || !instanceToken.componentNode) {
      continue; // Skip remote instances and instances without component references
    }

    try {
      // Get the actual Figma nodes for comparison
      const instanceNode = await figma.getNodeByIdAsync(instanceId) as InstanceNode;
      const componentNode = instanceToken.componentNode;
      
      if (!instanceNode || !componentNode) {
        continue;
      }

      // Extract fresh styles from both nodes for accurate comparison
      const componentSetToken = instanceToken.componentNode?.parent?.type === 'COMPONENT_SET' 
        ? collection.componentSets[instanceToken.componentNode.parent.id]
        : undefined;
      
      const componentTokenData = collection.components[componentNode.id];
      
      const instanceStyles = await extractInstanceStyles(instanceNode, componentTokenData, componentSetToken);
      const componentStyles = await extractComponentStyles(componentNode, componentTokenData, componentSetToken);

      // Find overrides by comparing properties
      const overrideTokens: StyleToken[] = [];
      
      for (const instanceStyle of instanceStyles) {
        const correspondingComponentStyle = componentStyles.find(
          componentStyle => componentStyle.property === instanceStyle.property
        );

        // If there's no corresponding token in the component, or values differ, it's an override
        if (!correspondingComponentStyle || 
            correspondingComponentStyle.value !== instanceStyle.value ||
            correspondingComponentStyle.rawValue !== instanceStyle.rawValue) {
          overrideTokens.push(instanceStyle);
        }
      }

      // Include this instance even if there are no overrides (as per requirement)
      overrides.push({
        instanceId,
        instanceName: instanceToken.name,
        componentId: componentNode.id,
        overrideTokens,
        allTokens: instanceStyles,
      });

    } catch (error) {
      console.warn(`Failed to process instance ${instanceToken.name}:`, error);
      // Still include the instance with empty overrides
      overrides.push({
        instanceId,
        instanceName: instanceToken.name,
        componentId: instanceToken.componentNode?.id,
        overrideTokens: [],
        allTokens: [],
      });
    }
  }

  return overrides;
}

/**
 * Extracts all style properties for an instance node by processing it directly.
 * This is used to get fresh token data for comparison.
 */
export async function extractInstanceStyles(
  instanceNode: InstanceNode,
  componentToken?: ComponentToken,
  componentSetToken?: any
): Promise<StyleToken[]> {
  const nodePathNames = getNodePathNames(instanceNode);
  const processors = getProcessorsForNode(instanceNode);
  const allTokens: StyleToken[] = [];

  for (const processor of processors) {
    const tokens = await extractNodeToken(
      instanceNode,
      processor,
      nodePathNames,
      componentToken,
      componentSetToken,
    );

    const styleTokens = tokens.filter((token): token is StyleToken => token.type === 'style');
    allTokens.push(...styleTokens);
  }

  return allTokens;
}

/**
 * Extracts all style properties for a component node by processing it directly.
 */
export async function extractComponentStyles(
  componentNode: ComponentNode,
  componentToken?: ComponentToken,
  componentSetToken?: any
): Promise<StyleToken[]> {
  const nodePathNames = getNodePathNames(componentNode);
  const processors = getProcessorsForNode(componentNode);
  const allTokens: StyleToken[] = [];

  for (const processor of processors) {
    const tokens = await extractNodeToken(
      componentNode,
      processor,
      nodePathNames,
      componentToken,
      componentSetToken,
    );

    const styleTokens = tokens.filter((token): token is StyleToken => token.type === 'style');
    allTokens.push(...styleTokens);
  }

  return allTokens;
}
