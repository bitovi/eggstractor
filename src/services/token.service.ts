import {
  ComponentSetToken,
  ComponentToken,
  InstanceToken,
  PathNode,
  StyleToken,
  VariableToken,
} from '../types';
import { StyleProcessor, VariableBindings } from '../types/processors';
import { collectBoundVariable } from './variable.service';

export const extractInstanceSetToken = async (node: InstanceNode): Promise<InstanceToken> => {
  const componentNode = await node.getMainComponentAsync();

  return {
    type: node.type,
    id: node.id,
    name: node.name,
    remote: componentNode?.remote ?? false,
    // This can be used to reference components -> component sets for non-remote components
    componentNode,
    variantProperties: node.variantProperties ?? {},
  };
};

export const extractComponentToken = (
  node: ComponentNode,
  componentSetToken?: ComponentSetToken,
): ComponentToken => {
  return {
    type: node.type,
    id: node.id,
    componentSetId: componentSetToken?.id ?? null,
    variantProperties: node.variantProperties ?? {},
  };
};

export const extractComponentSetToken = (node: ComponentSetNode): ComponentSetToken => {
  const variantPropertyDefinitions: Record<string, string[]> = {};

  for (const [key, value] of Object.entries(node.componentPropertyDefinitions)) {
    if (value.type !== 'VARIANT') {
      continue;
    }

    variantPropertyDefinitions[key] = value.variantOptions ?? [];
  }

  return {
    type: node.type,
    id: node.id,
    name: node.name,
    variantPropertyDefinitions,
  };
};

const variableTokensCache = new Map<string, VariableToken>();

export async function extractNodeToken(
  node: SceneNode,
  processor: StyleProcessor,
  pathNodes: PathNode[],
  componentToken?: ComponentToken | null,
  componentSetToken?: ComponentSetToken | null,
): Promise<(StyleToken | VariableToken)[]> {
  const tokens: (StyleToken | VariableToken)[] = [];
  const variableTokensMap = new Map<string, VariableToken>();

  // Helper to check or add variable token
  const getOrCreateVariableToken = async (varId: string, property: string) => {
    const key = `${varId}-${property}`;

    let token: VariableToken | null | undefined;

    // Check global cache first
    if (variableTokensCache.has(key)) {
      token = variableTokensCache.get(key)!;
    }

    // All variable tokens of a property should already be unique and varId
    // shouldn't have to be included. This is because any style token should
    // only attempt to set one css property.
    token ??= await collectBoundVariable(varId, property, pathNodes, node);

    if (!token) {
      throw new Error('Unexpected null token for variable');
    }

    // Check scoped cache to StyleToken
    if (!variableTokensMap.has(key)) {
      variableTokensMap.set(key, token);
      // Deprecated and should be removed in the future
      tokens.push(token);
    }

    return token;
  };

  // Step 1 & 2: Handle Variable Bindings
  const customBoundVariables = node.boundVariables as unknown as VariableBindings;
  const bindings = processor.bindingKey
    ? Array.isArray(customBoundVariables[processor.bindingKey])
      ? (customBoundVariables[processor.bindingKey] as VariableAlias[])
      : ([customBoundVariables[processor.bindingKey]] as VariableAlias[])
    : [];

  for (const binding of bindings) {
    if (binding?.id) {
      await getOrCreateVariableToken(binding.id, processor.property);
    }
  }

  // Step 3: Collect variables from boundVariables
  if ('boundVariables' in node && node.boundVariables) {
    for (const [key, value] of Object.entries(node.boundVariables)) {
      if (typeof key === 'string' && value) {
        if (Array.isArray(value)) {
          for (const v of value) {
            if (v.type === 'VARIABLE_ALIAS') {
              await getOrCreateVariableToken(v.id, key);
            }
          }
        } else if (value?.type === 'VARIABLE_ALIAS') {
          await getOrCreateVariableToken(String(value.id), key);
        }
      }
    }
  }

  const variableTokenMapByProperty = new Map<string /* property */, VariableToken>();

  variableTokensMap.forEach((token, key) => {
    if (!variableTokenMapByProperty.has(token.property)) {
      variableTokenMapByProperty.set(token.property, token);
    } else {
      // All VariableTokens for a single StyleToken should be unique by property
      // This is because we currently are not supporting multiple variable
      // tokens for the same property.
      throw new Error(
        `Variable token for property "${token.property}" already exists in the map. Key: ${key}`,
      );
    }
  });

  // Step 4: Process the node and create Style Token
  const processedValue = await processor.process(variableTokenMapByProperty, node);

  if (processedValue) {
    const styleToken: StyleToken = {
      type: 'style',
      name: pathNodes.map(({ name }) => name).join('_'),
      value: processedValue.value,
      rawValue: processedValue.rawValue,
      valueType: processedValue.valueType,
      property: processor.property,
      path: pathNodes,
      // Deprecated and will be removed in the future
      variables: variableTokensMap.size > 0 ? [...variableTokensMap.values()] : undefined,
      variableTokenMapByProperty,
      metadata: {
        figmaId: node.id,
      },
      warnings: processedValue.warnings,
      errors: processedValue.errors,
      componentId: componentToken?.id,
      componentSetId: componentSetToken?.id,
    };
    tokens.push(styleToken);
  }

  return tokens;
}
