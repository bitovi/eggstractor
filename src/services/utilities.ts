/**
 * Used for showing progress in the UI. It is capped at 95% during the time that
 * nodes are being processed to tokens. The last 5% is reserved for
 * transforming the tokens to the requested format.
 */
export const MAX_PROGRESS_PERCENTAGE = 95;

/**
 * Promise wrapped setTimeout.
 */
export function delay(ms: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, ms);
  });
}
