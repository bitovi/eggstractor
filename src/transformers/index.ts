export * from "./scss.transformer";
export * from "./css.transformer";
