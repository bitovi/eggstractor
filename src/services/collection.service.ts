import { ComponentSetToken, ComponentToken, TokenCollection } from '../types';
import { getProcessorsForNode } from '../processors';
import { extractComponentSetToken, extractComponentToken, extractNodeToken } from '../services';
import { getNodePathNames } from '../utils/node.utils';

function getFlattenedValidNodes(node: BaseNode): BaseNode[] {
  const result: BaseNode[] = [];

  function traverse(currentNode: BaseNode) {
    // Skip VECTOR and INSTANCE nodes entirely. INSTANCE nodes which are
    // instances of components, which are not relevant for token extraction.
    if ('type' in currentNode && ['VECTOR', 'INSTANCE'].includes(currentNode.type)) {
      return;
    }

    result.push(currentNode);

    if ('children' in currentNode) {
      for (const child of currentNode.children) {
        traverse(child);
      }
    }
  }

  traverse(node);
  return result;
}

export async function collectTokens(onProgress: (progress: number, message: string) => void) {
  const collection: TokenCollection = { tokens: [], components: {}, componentSets: {} };

  let componentToken: ComponentToken | null = null;
  let componentSetToken: ComponentSetToken | null = null;

  let totalNodes = 0;
  let processedNodes = 0;

  let lastTimestamp = Date.now();
  let lastPercentage = -1;

  async function processNode(node: BaseNode) {
    processedNodes++;

    // calculate integer percent in your 10–95 range
    const currentPercentage = Math.floor((processedNodes / totalNodes) * 85) + 10;
    const shouldUpdate = currentPercentage !== lastPercentage || processedNodes === totalNodes;

    if (shouldUpdate) {
      lastTimestamp = currentPercentage;
      onProgress(currentPercentage, `Processing nodes… ${processedNodes}/${totalNodes}`);

      // throttle yields to at most once every 200 ms
      const now = Date.now();
      if (now - lastTimestamp >= 200) {
        await new Promise((r) => setTimeout(r, 0));
        lastTimestamp = now;
      }
    }

    if ('type' in node && 'boundVariables' in node) {
      if (node.type === 'COMPONENT_SET') {
        componentSetToken = extractComponentSetToken(node);
        collection.componentSets[node.id] = componentSetToken;
      }

      if (node.type === 'COMPONENT') {
        componentToken = extractComponentToken(node, componentSetToken!);
        collection.components[node.id] = componentToken;
      }

      const nodePathNames = getNodePathNames(node);
      const processors = getProcessorsForNode(node);

      for (const processor of processors) {
        const tokens = await extractNodeToken(
          node,
          processor,
          nodePathNames,
          componentToken,
          componentSetToken,
        );
        collection.tokens.push(...tokens);
      }
    }
  }

  onProgress(0, 'Loading pages...');
  await figma.loadAllPagesAsync();

  onProgress(5, 'Counting nodes...');
  const validNodes = figma.root.children.flatMap((page) => getFlattenedValidNodes(page));
  totalNodes = validNodes.length;

  onProgress(10, `Processing ${totalNodes} nodes...`);

  // Process all pages in parallel for maximum speed
  const pagePromises = figma.root.children.map(async (page) => {
    const validNodes = getFlattenedValidNodes(page);

    for (const node of validNodes) {
      await processNode(node);
    }
  });

  await Promise.all(pagePromises);

  return collection as Readonly<TokenCollection>;
}
