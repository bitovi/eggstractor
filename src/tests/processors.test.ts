import { collectTokens } from '../services';
import { transformToCss } from '../transformers';
import testData from './fixtures/figma-test-data.json';
import testDataSimple from './fixtures/figma-test-simple.json';

// Add this helper function at the top of the test file
function parseCssClass(css: string, className: string): string | null {
  const regex = new RegExp(`\\.${className}\\s*{([^}]*)}`, 'g');
  const match = regex.exec(css);
  if (!match) return null;
  return match[1].trim();
}

describe.skip('Processors', () => {
  it('should process background gradient correctly', async () => {    
    // Create proper node hierarchy in test data
    const pageNode = {
      ...testData,
      type: 'PAGE',
      name: 'background',
      parent: null,
      width: 100,  // Add default dimensions
      height: 100
    };

    // Update children to have proper parent references
    const children = testData.children.map((child: BaseNode) => ({
      ...child,
      parent: pageNode,
      width: 100,  // Add dimensions to children
      height: 100
    }));
    pageNode.children = children;

    // Mock minimal Figma API with proper variable structure
    global.figma = {
      currentPage: pageNode,
      variables: {
        getVariableByIdAsync: async (id: string) => {
          const variable = testData.variables[id as keyof typeof testData.variables];
          if (!variable) return null;

          return Promise.resolve({
            ...variable,
            description: "",
            hiddenFromPublishing: false,
            resolvedType: "COLOR",
            scopes: ["ALL_SCOPES"],
            codeSyntax: {},
            remote: false,
            documentation: "",
            getPublishStatusAsync: async () => ({ status: "UNPUBLISHED" }),
            setValueForMode: () => {},
            remove: () => {},
            removeValueForMode: () => {},
            exposedToLibrary: false,
            subscribedLibraryId: null,
            subscribedLibraryOriginId: null,
            isCollectionPublished: false,
            boundVariables: [],
            variableCollectionId: "1",
            key: "key",
            resolveForConsumer: () => null,
            setVariableCodeSyntax: () => {},
            setName: () => {},
            setDescription: () => {},
            setExposedToLibrary: () => {},
            setHiddenFromPublishing: () => {},
            setScopes: () => {},
            setResolvedType: () => {},
            removeVariableCodeSyntax: () => {},
            getPluginData: () => "",
            setPluginData: () => {},
            getPluginDataKeys: () => [],
            getSharedPluginData: () => "",
            setSharedPluginData: () => {},
            removed: false
          } as unknown as Variable);
        }
      }
    };

    const tokens = await collectTokens();    
    const css = transformToCss(tokens);

    // Test specific styles with snapshots
    const styles = {
      linear: parseCssClass(css, 'background-gradient-linear-style'),
      radial: parseCssClass(css, 'background-gradient-radial-style'),
      angular: parseCssClass(css, 'background-gradient-angular-style'),
      linearAlpha: parseCssClass(css, 'background-gradient-linear-alpha-style')
    };

    // Snapshot all styles
    expect(styles).toMatchSnapshot('gradient styles');

    // Keep direct assertions for critical values
    expect(styles.linear).toBe('background: linear-gradient(224deg, #00464A 3.74%, #04646A 98.29%);');
    expect(styles.linearAlpha).toBe('background:  linear-gradient(224deg, rgba(0, 70, 74, 0.50) 3.74%, #04646A 98.29%);');
  });

  it('should process background gradient simple correctly', async () => {    
    // Create proper node hierarchy in test data
    const pageNode = {
      ...testDataSimple,
      type: 'PAGE',
      name: 'background',
      parent: null,
      width: 100,  // Add default dimensions
      height: 100
    };

    // Update children to have proper parent references
    const children = testDataSimple.children.map((child: BaseNode) => ({
      ...child,
      parent: pageNode,
      width: 100,  // Add dimensions to children
      height: 100
    }));
    pageNode.children = children;

    // Mock minimal Figma API with proper variable structure
    global.figma = {
      currentPage: pageNode,
      variables: {
        getVariableByIdAsync: async (id: string) => {
          const variable = testDataSimple.variables[id as keyof typeof testDataSimple.variables];
          if (!variable) return null;

          return Promise.resolve({
            ...variable,
            description: "",
            hiddenFromPublishing: false,
            resolvedType: "COLOR",
            scopes: ["ALL_SCOPES"],
            codeSyntax: {},
            remote: false,
            documentation: "",
            getPublishStatusAsync: async () => ({ status: "UNPUBLISHED" }),
            setValueForMode: () => {},
            remove: () => {},
            removeValueForMode: () => {},
            exposedToLibrary: false,
            subscribedLibraryId: null,
            subscribedLibraryOriginId: null,
            isCollectionPublished: false,
            boundVariables: [],
            variableCollectionId: "1",
            key: "key",
            resolveForConsumer: () => null,
            setVariableCodeSyntax: () => {},
            setName: () => {},
            setDescription: () => {},
            setExposedToLibrary: () => {},
            setHiddenFromPublishing: () => {},
            setScopes: () => {},
            setResolvedType: () => {},
            removeVariableCodeSyntax: () => {},
            getPluginData: () => "",
            setPluginData: () => {},
            getPluginDataKeys: () => [],
            getSharedPluginData: () => "",
            setSharedPluginData: () => {},
            removed: false
          } as unknown as Variable);
        }
      }
    };

    const tokens = await collectTokens();    
    const css = transformToCss(tokens);

    // Test specific styles with snapshots
    const styles = {
      linear: parseCssClass(css, 'background-gradient-top-bottom'),
    };

    // Keep direct assertions for critical values
    expect(styles.linear).toBe('background: linear-gradient(180deg, #000 -6.5%, #FFF 110%);');
  });
}); 