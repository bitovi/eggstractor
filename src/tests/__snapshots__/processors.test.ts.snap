// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processors should process background gradient correctly: gradient styles 1`] = `
{
  "angular": "background: conic-gradient(from 134deg at 50% 50%, #00464A 0deg, #04646A 360deg);",
  "linear": "background: linear-gradient(224deg, #00464A 3.74%, #04646A 98.29%);",
  "linearAlpha": "background:  linear-gradient(224deg, rgba(0, 70, 74, 0.50) 3.74%, #04646A 98.29%);",
  "radial": "background: radial-gradient(100% 100% at 100% 100%, #00464A 0%, #04646A 100%);",
}
`;
